src/three-components/PMREMGenerator.ts

```diff
@@ -0,0 +1,367 @@
+    for (let k = index * 6; k < (index + 1) * 6; k++) {
+      const material = (this.objects[k].material as any);
+      material.uniforms['envMap'].value = target.texture;
+      material.envMap = target.texture;
+    }
+  }
+
+  generateMipmaps(cubeTarget: WebGLRenderTargetCube) {
+    const gammaInput = this.renderer.gammaInput;
+    const gammaOutput = this.renderer.gammaOutput;
+    const toneMapping = this.renderer.toneMapping;
+    const toneMappingExposure = this.renderer.toneMappingExposure;
+    const currentRenderTarget = this.renderer.getRenderTarget();
+
+    this.renderer.toneMapping = LinearToneMapping;
+    this.renderer.toneMappingExposure = 1.0;
+    this.renderer.gammaInput = false;
+    this.renderer.gammaOutput = false;
+
+    this.mipmapShader.uniforms['texelSize'].value = 1.0 / cubeTarget.width;
+    this.mipmapShader.uniforms['envMap'].value = cubeTarget.texture;
+    (this.mipmapShader as any).envMap = cubeTarget.texture;
+    for (let i = this.numLods - 1; i >= 0; i--) {
+      this.cubeCamera.renderTarget = this.cubeLods[i];
+      this.cubeCamera.update(this.renderer, this.mipmapScene);
+      this.mipmapShader.uniforms['texelSize'].value =
+          1.0 / this.cubeLods[i].width;
+      this.mipmapShader.uniforms['envMap'].value = this.cubeLods[i].texture;
+      (this.mipmapShader as any).envMap = this.cubeLods[i].texture;
+    }
+
+    this.renderer.setRenderTarget(currentRenderTarget);
+    this.renderer.toneMapping = toneMapping;
+    this.renderer.toneMappingExposure = toneMappingExposure;
+    this.renderer.gammaInput = gammaInput;
+    this.renderer.gammaOutput = gammaOutput;
+  }
+
+  packMipmaps() {
+    this.objects.forEach((object) => {
+      this.packingScene.add(object);
+    });
+
+    const gammaInput = this.renderer.gammaInput;
+    const gammaOutput = this.renderer.gammaOutput;
+    const toneMapping = this.renderer.toneMapping;
+    const toneMappingExposure = this.renderer.toneMappingExposure;
+    const currentRenderTarget = this.renderer.getRenderTarget();
+
+    this.renderer.gammaInput = false;
+    this.renderer.gammaOutput = false;
+    this.renderer.toneMapping = LinearToneMapping;
+    this.renderer.toneMappingExposure = 1.0;
+    this.renderer.setRenderTarget(this.cubeUVRenderTarget);
+    this.renderer.render(this.packingScene, this.flatCamera);
+
+    this.renderer.setRenderTarget(currentRenderTarget);
+    this.renderer.toneMapping = toneMapping;
+    this.renderer.toneMappingExposure = toneMappingExposure;
+    this.renderer.gammaInput = gammaInput;
+    this.renderer.gammaOutput = gammaOutput;
+
+    this.objects.forEach((object) => {
+      this.packingScene.remove(object);
+    });
+    this.cubeLods.forEach((target) => {
+      target.dispose();
+    });
+    this.cubeLods = [];
+  }
+
+  getMipmapShader() {
+    var shaderMaterial = new ShaderMaterial({
+
+      uniforms: {'texelSize': {value: 0.5}, 'envMap': {value: null}},
```

Comment:
No need to use quoted property key names here. It's only necessary if you are using characters that are not typically allowed as part of a JavaScript identifier. For example:

```typescript
const obj = { my-key: 0 }; // Syntax error, because the hyphen is not allowed

const obj = { 'my-key': 0 }; // Allowed, because the key name has been quoted
```